pkg/resource/deploy/step_executor.go

```diff
@@ -0,0 +1,301 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploy
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/pkg/diag"
+	"github.com/pulumi/pulumi/pkg/util/contract"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+)
+
+const (
+	// Dummy workerID for synchronous operations.
+	synchronousWorkerID = -1
+
+	// Utility constant for easy debugging.
+	stepExecutorLogLevel = 4
+)
+
+var (
+	// errStepApplyFailed is a sentinel error for errors that arise when step application fails.
+	// We (the step executor) are not responsible for reporting those errors so this sentinel ensures
+	// that we don't do so.
+	errStepApplyFailed = errors.New("step application failed")
+)
+
+// A Chain is a sequence of Steps that must be executed in the given order.
+type Chain = []Step
+
+// stepExecutor is the component of the engine responsible for taking steps and executing
+// them, possibly in parallel if requested. The step generator operates on the granularity
+// of "chains", which are sequences of steps that must be executed exactly in the given order.
+// Chains are a simplification of the full dependency graph DAG within Pulumi programs. Since
+// Pulumi language hosts can only invoke the resource monitor once all of their dependencies have
+// resolved, we (the engine) can assume that any chain given to us by the step generator is already
+// ready to execute.
+type stepExecutor struct {
```

Comment:
I'm still not convinced that we need separate types here given that the step executor and the plan executor share nearly all of their state. Non-blocking, but I do think that this adds some unnecessary complexity.

Reply:
My original design for these two components accommodated a full-blown DAG execution engine. It made a lot more sense to separate these two components when the step executor was executing DAGs instead of just iterating over lists of steps. We might still need a DAG execution engine, so I'm thinking we can always merge the two structures if it turns out it's not useful to separate them once the dust has settled.